### PR TITLE
Feature/main menu

### PR DESCRIPTION
**Pull request by samuelWilliams99**
image
image
image
image

### CODE REVIEW

File: gamemode/init.lua [hunk 1]
```diff
@@ -6,7 +6,8 @@ resource.AddFile( "resource/fonts/bungee_regular.ttf" )
 
 -- Needed to show holding animations, to be remove when main menu implemented
 function GM:PlayerSetModel( ply )
-    ply:SetModel( "models/player/odessa.mdl" )
+    ply.PlayerModel = ply.PlayerModel or "models/player/odessa.mdl"
```
Comment (sarahsturgeon) [inline gamemode/init.lua]:
```suggestion
    ply.PlayerModel = ply.PlayerModel or RVR.Config.DefaultPlayermodel
```
?

File: gamemode/modules/main_menu/cl_mainmenu.lua [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    font = "Bungee Regular",
+    size = ScrH() * 0.05,
+    weight = 700,
+} )
+
+surface.CreateFont( "RVR_StartMenuTextEntry", {
+    font = "Roboto",
+    size = ScrH() * 0.04,
+    weight = 700,
+} )
```
Comment (sarahsturgeon) [inline gamemode/modules/main_menu/cl_mainmenu.lua]: Might be nice to pull these out into a `fonts.lua`

File: gamemode/modules/main_menu/cl_mainmenu_create.lua [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        self:SetPos( _x + _w + 20, _y )
+    end
+
+    local entry = vgui.Create( entryClass, container )
+    entry:SetPos( x + 10, y )
+    entry:SetSize( entryW, entryH )
+    if tooltip then entry:SetTooltip( tooltip ) end
+
+    return entry
+end
```
Comment (sarahsturgeon) [inline gamemode/modules/main_menu/cl_mainmenu_create.lua]: Some magic numbers in here, some repeated, would be nice to get those into variables

(actually, applies to the whole file)

**Review by sarahsturgeon — CHANGES_REQUESTED** [review]
Almost all the CL code in here has a lot of magic numbers, many of them repeated

I think it would be prudent to pull those out into named vars

File: gamemode/modules/main_menu/cl_mainmenu.lua (hunk 2 above)
Comment (samuelWilliams99) [reply gamemode/modules/main_menu/cl_mainmenu.lua]: Will do in the future, project wide. Made a card

File: gamemode/modules/main_menu/cl_mainmenu_create.lua (hunk 3 above)
Comment (overlordbleck) [reply gamemode/modules/main_menu/cl_mainmenu_create.lua]: Oh god no

Comment (samuelWilliams99) [reply gamemode/modules/main_menu/cl_mainmenu_create.lua]: Made some changes, what do you think?

**Review by sarahsturgeon — APPROVED** [review]

File: gamemode/config/generic.lua [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    "models/player/group03m/female_04.mdl",
+    "models/player/group03m/female_05.mdl",
+    "models/player/group03m/female_06.mdl",
+    "models/player/monk.mdl",
+    "models/player/mossman.mdl",
+    "models/player/mossman_arctic.mdl",
+    "models/player/group02/male_02.mdl",
+    "models/player/group02/male_04.mdl",
+    "models/player/group02/male_06.mdl",
+    "models/player/group02/male_08.mdl",
```
Comment (overlordbleck) [inline gamemode/config/generic.lua]:
```suggestion
    "models/player/group02/male_08.mdl"
```

File: gamemode/modules/main_menu/cl_mainmenu.lua [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    elseif page == mainMenu.MENU_JOIN then
+        mainMenu.createPartyJoinMenu( ... )
+    elseif page == mainMenu.MENU_MODEL then
+        mainMenu.createModelMenu( ... )
+    end
+end
+
+local function lerpCol( a, b, l )
+    return Color( Lerp( l, a.r, b.r ), Lerp( l, a.g, b.g ), Lerp( l, a.b, b.b ), Lerp( l, a.a, b.a ) )
+end
```
Comment (overlordbleck) [inline gamemode/modules/main_menu/cl_mainmenu.lua]: too short variable names

File: gamemode/modules/main_menu/cl_mainmenu_create.lua [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    function errorLabel:PerformLayout()
+        local _w = label:GetWide()
+        local _x, _y = label:GetPos()
+        self:SetPos( _x + _w + 20, _y )
+    end
+
+    local entry = vgui.Create( entryClass, container )
+    entry:SetPos( x + 10, y )
+    entry:SetSize( entryW, entryH )
+    if tooltip then entry:SetTooltip( tooltip ) end
```
Comment (overlordbleck) [inline gamemode/modules/main_menu/cl_mainmenu_create.lua]:
```suggestion
    if tooltip then
        entry:SetTooltip( tooltip )
    end
```

File: gamemode/modules/main_menu/cl_mainmenu_join.lua [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        table.insert( parties, RVR.Party.parties[key] )
+    end
+
+    mainMenu.fillPartiesList( partyBox, parties )
+
+    mainMenu.createNormalButton( "BACK", 50, mainMenu.h - 40 - ScrH() * 0.04, TEXT_ALIGN_LEFT, function()
+        if mainMenu.awaitingReply then return end
+        mainMenu.setMenuPage( mainMenu.MENU_START )
+    end )
+end
```
Comment (overlordbleck) [inline gamemode/modules/main_menu/cl_mainmenu_join.lua]:
```suggestion
end

```

File: gamemode/modules/main_menu/cl_mainmenu_create.lua [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+                end
+
+                mainMenu.awaitingReply = false
+            end
+        end )
+
+        mainMenu.awaitingReply = true
+    end )
+end
+
```
Comment (overlordbleck) [inline gamemode/modules/main_menu/cl_mainmenu_create.lua]:
```suggestion
```

**Review by overlordbleck — APPROVED** [review]

File: gamemode/config/generic.lua [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    "models/player/group03m/female_04.mdl",
+    "models/player/group03m/female_05.mdl",
+    "models/player/group03m/female_06.mdl",
+    "models/player/monk.mdl",
+    "models/player/mossman.mdl",
+    "models/player/mossman_arctic.mdl",
+    "models/player/group02/male_02.mdl",
+    "models/player/group02/male_04.mdl",
+    "models/player/group02/male_06.mdl",
+    "models/player/group02/male_08.mdl"
```
Comment (plally) [inline gamemode/config/generic.lua:72]: hmmmm

Comment (samuelWilliams99) [reply gamemode/config/generic.lua:72]: Can't please everyone

File: gamemode/init.lua [hunk 10]
```diff
@@ -6,7 +6,8 @@ resource.AddFile( "resource/fonts/bungee_regular.ttf" )
 
 -- Needed to show holding animations, to be remove when main menu implemented
 function GM:PlayerSetModel( ply )
-    ply:SetModel( "models/player/odessa.mdl" )
+    ply.PlayerModel = ply.PlayerModel or RVR.Config.Generic.DEFAULT_PLAYER_MODEL
```
Comment (plally) [inline gamemode/init.lua]: camelCase

File: gamemode/modules/main_menu/cl_mainmenu.lua [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+} )
+
+hook.Add( "HUDShouldDraw", "RVR_MainMenu_HideHud", function( hudType )
+    if hudType == "CHudGMod" then return end
+    if mainMenu.frame or mainMenu.hideGui then return false end
+end )
+
+hook.Add( "HUDPaintBackground", "RVR_MainMenu", function()
+    if not mainMenu.frame then return end
+    local _w, _h = ScrW(), ScrH()
```
Comment (plally) [inline gamemode/modules/main_menu/cl_mainmenu.lua]: why _w _h?

**Review by plally — CHANGES_REQUESTED** [review]

File: gamemode/modules/main_menu/cl_mainmenu_create.lua [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+            draw.RoundedBox( 10, 0, 0, _w, _h, bgCol )
+        end
+
+        btn:SetSelected( k == 1 )
+        table.insert( btns, btn )
+    end
+end
+
+local joinModeLookup = {
+    ["PUBLIC"] = 0,
```
Comment (plally) [inline gamemode/modules/main_menu/cl_mainmenu_create.lua:240]: would defining  the table like `PUBLIC = 0` be better

File: gamemode/modules/main_menu/cl_mainmenu_create.lua [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+    mainMenu.createErrorLabel( "generic", mainMenu.w * 0.5, mainMenu.h * 0.94, true )
+
+    mainMenu.awaitingReply = false
+
+    mainMenu.createNormalButton( "BACK", 50, mainMenu.h - 40 - ScrH() * 0.04, TEXT_ALIGN_LEFT, function()
+        if mainMenu.awaitingReply then return end
+        mainMenu.setMenuPage( mainMenu.MENU_START )
+    end )
+
+    mainMenu.createNormalButton( "CREATE", mainMenu.w - 50, mainMenu.h - 40 - ScrH() * 0.04, TEXT_ALIGN_RIGHT, function()
```
Comment (plally) [inline gamemode/modules/main_menu/cl_mainmenu_create.lua]: this pretty long, at least 117 characters

File: gamemode/modules/main_menu/cl_mainmenu_create.lua (hunk 12 above)
Comment (samuelWilliams99) [reply gamemode/modules/main_menu/cl_mainmenu_create.lua:240]: no

File: gamemode/modules/main_menu/cl_mainmenu.lua (hunk 11 above)
Comment (samuelWilliams99) [reply gamemode/modules/main_menu/cl_mainmenu.lua]: Fixed

**Pull request merged**